src/verifast1.ml

```diff
@@ -2785,6 +2841,96 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
       | Int (Unsigned, n1), Int (Signed, n2) -> if n2 <= n1 then t1 else t2
       | Int (s, n1), Int (_, n2) -> Int (s, max n1 n2)
 
+  open Java_Inference
+
+  let rec super_types (ObjType (cn, targs)) =
+    let supers =
+      if cn = "java.lang.Object" then
+        []
+      else
+        direct_supers (ObjType (cn, targs))
+    in
+    flatmap (fun s -> s::super_types s) supers
+
+  (* calculates the shared supertype that is more specific than any other shared supertype *)
+  let least_upper_bound l tps =
+    match tps with
+      [t] -> [t]
+    | [] -> failwith "calculating least upper bound of no type"
+    | tps -> 
+      let st_map = List.map (fun tp -> (tp, tp::super_types tp)) tps in
+      let filtered_candidate_set = (*intersection of all est sets *)
+        let intersection setA setB = List.filter (fun memA -> List.mem memA setB) setA in
+        let rec iter sets iset = match sets with
+          [] -> []
+        | [last] -> intersection iset last
+        | hd::tl -> iter tl (intersection iset hd)
+        in
+        iter (List.map snd st_map) (snd (List.hd st_map))
+      in
+      (* MEC = { V | V in EC, and for all W ≠ V in EC, it is NOT the case that W <: V } *)
+      let mec = filtered_candidate_set |> List.filter begin fun v -> (* V in EC *)
+        let children = filtered_candidate_set |> List.filter begin fun w ->  (* for every V, select all it's children in EC that are not V *)
+          if (w = v) then   
+            false
+          else
+            (is_subtype_of_ w v)
+        end
+        in (* if V has no children, it is the most specific candidate*)
+        children = []
+      end
+      in
+      mec 
+
+  let infer_type l unknownTypes knownTypes =
+    let rec inference_set set actual expected = 

```

Comment:
(Neither `unknownTypes` and `knownTypes` nor `actual` and `expected` seem like good names to me.)